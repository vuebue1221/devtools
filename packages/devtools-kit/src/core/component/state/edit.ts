export interface EditStatePayload<T = unknown> {
  value: T
  newKey: string | null
  remove?: false
}
