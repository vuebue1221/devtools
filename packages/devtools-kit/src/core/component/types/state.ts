import type { VueAppInstance } from '@vue-devtools-next/schema'

export interface InspectorCustomState {
  _custom?: {
    type?: string
    displayText?: string
    tooltipText?: string
    value?: string
    stateTypeName?: string
    fields?: {
      abstract?: boolean
    }
  }
}

export interface InspectorState {
  key: string
  value: string | number | Record<string, unknown> | InspectorCustomState | Array<unknown>
  type: string
  stateTypeName?: string
  meta?: Record<string, boolean | string>
  raw?: string
  editable?: boolean
  children?: {
    key: string
    value: string | number
    type: string
  }[]
}

export interface InspectorStateApiPayload {
  app: VueAppInstance
  inspectorId: string
  nodeId: string
}

export interface AddInspectorApiPayload {
  id: string
  label: string
  icon?: string
  treeFilterPlaceholder?: string
  actions?: {
    icon: string
    tooltip: string
    action: (payload: unknown) => void
  }[]
}
