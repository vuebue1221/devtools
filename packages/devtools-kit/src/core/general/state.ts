import type { AppRecord } from '@vue-devtools-next/schema'
import { target as global } from '@vue-devtools-next/shared'
import type { DevToolsPluginApi } from '../../api'
import { DevToolsEvents, apiHooks } from '../../api'
import type { Router, RouterInfo } from '../router'
import { RouterKey, devtoolsRouterInfo, normalizeRouterInfo } from '../router'

const StateKey = '__VUE_DEVTOOLS_GLOBAL_STATE__'
const ContextKey = '__VUE_DEVTOOLS_CONTEXT__'
global[StateKey] ??= {
  connected: false,
  appRecords: [],
  activeAppRecord: null,
  selectedComponentId: null,
  pluginBuffer: [],
  tabs: [],
}
global[ContextKey] ??= {
  appRecord: null,
  api: null,
  inspector: [],
  timelineLayer: [],
  routerInfo: {},
  router: null,
  activeInspectorTreeId: '',
}

export const devtoolsState = new Proxy(global[StateKey], {
  get(target, property) {
    return global[StateKey][property]
  },
  set(target, property, value) {
    const oldState = global[StateKey]
    target[property] = value
    // sync to global to ensure the state is consistent
    global[StateKey][property] = value
    if (property === 'activeAppRecord') {
      // update context
      global[ContextKey].appRecord = value
      global[ContextKey].api = value.api
      global[ContextKey].inspector = value.inspector ?? []
      normalizeRouterInfo(value)
      global[ContextKey].routerInfo = devtoolsRouterInfo
    }

    apiHooks.callHook(DevToolsEvents.DEVTOOLS_STATE_UPDATED, global[StateKey], oldState)
    return true
  },
  deleteProperty(target, property) {
    delete target[property]
    return true
  },
})

Object.defineProperty(devtoolsState.tabs, 'push', {
  configurable: true,
  value(...items) {
    const result = Array.prototype.push.apply(this, items)
    devtoolsState.tabs = this
    apiHooks.callHook(DevToolsEvents.CUSTOM_TABS_UPDATED, this)
    return result
  },
})

export const devtoolsContext = new Proxy(global[ContextKey], {
  get(target, property) {
    if (property === 'router')
      return global[RouterKey]

    return global[ContextKey][property]
  },
}) as unknown as {
  appRecord: AppRecord
  api: DevToolsPluginApi
  inspector: {
    id: string
    nodeId: string
    filter: string
    treeFilterPlaceholder: string
  }[]
  timelineLayer: {
    id: string
    label: string
    color: number
  }[]
  routerInfo: RouterInfo
  router: Router
  activeInspectorTreeId: string
}
