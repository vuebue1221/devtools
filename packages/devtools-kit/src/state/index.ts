export * from './state'
export * from './app-record'
export * from './context'
export * from './router'
export * from './env'
