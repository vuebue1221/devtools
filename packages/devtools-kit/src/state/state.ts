import { target as global } from '@vue/devtools-shared'
import { debounce } from 'perfect-debounce'
import type { DevToolsState } from '../types'
import { DevToolsEvents, apiHooks } from '../api'

export type { DevToolsState } from '../types'

const STATE_KEY = '__VUE_DEVTOOLS_GLOBAL_STATE__'
const INITIAL_STATE = {
  connected: false,
  clientConnected: false,
  appRecords: [],
  activeAppRecord: null,
  selectedComponentId: null,
  pluginBuffer: [],
  tabs: [],
  commands: [],
  vitePluginDetected: false,
  activeAppRecordId: null,
}

global[STATE_KEY] ??= INITIAL_STATE

export function resetDevToolsState() {
  global[STATE_KEY] = INITIAL_STATE
}

export const callStateUpdatedHook = debounce((state: DevToolsState, oldState: DevToolsState) => {
  apiHooks.callHook(DevToolsEvents.DEVTOOLS_STATE_UPDATED, state, oldState)
}, 80)

export const callConnectedUpdatedHook = debounce((state: DevToolsState, oldState: DevToolsState) => {
  apiHooks.callHook(DevToolsEvents.DEVTOOLS_CONNECTED_UPDATED, state, oldState)
}, 80)

export const devtoolsState: DevToolsState = new Proxy(global[STATE_KEY], {
  get(target, property) {
    return global[STATE_KEY][property]
  },
  deleteProperty(target, property) {
    delete target[property]
    return true
  },
  set(target, property, value) {
    const oldState = { ...global[STATE_KEY] }

    target[property] = value
    // sync to global to ensure the state is consistent
    global[STATE_KEY][property] = value

    callStateUpdatedHook(global[STATE_KEY], oldState)
    if (['connected', 'clientConnected'].includes(property.toString()) && oldState[property] !== value)
      callConnectedUpdatedHook(global[STATE_KEY], oldState)

    return true
  },
})

Object.defineProperty(devtoolsState.tabs, 'push', {
  configurable: true,
  value(...items: unknown[]) {
    const result = Array.prototype.push.apply(this, items)
    devtoolsState.tabs = this
    apiHooks.callHook(DevToolsEvents.CUSTOM_TABS_UPDATED, this)
    return result
  },
})

;['push', 'splice'].forEach((method) => {
  Object.defineProperty(devtoolsState.commands, method, {
    configurable: true,
    value(...args: unknown[]) {
      const result = Array.prototype[method].apply(this, args)
      devtoolsState.commands = this
      apiHooks.callHook(DevToolsEvents.CUSTOM_COMMANDS_UPDATED, this)
      return result
    },
  })
})
