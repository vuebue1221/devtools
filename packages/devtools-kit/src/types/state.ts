import type { CustomCommand } from '../core/custom-command'
import type { CustomTab } from '../core/custom-tab'
import type { AppRecord, PluginDescriptor, PluginSetupFunction } from './app'

export interface DevToolsState {
  connected: boolean
  clientConnected: boolean
  vitePluginDetected: boolean
  appRecords: AppRecord[]
  activeAppRecord: AppRecord | null
  selectedComponentId: string | null
  pluginBuffer: [PluginDescriptor, PluginSetupFunction][]
  tabs: CustomTab[]
  commands: CustomCommand[]
  activeAppRecordId: string | null
}
