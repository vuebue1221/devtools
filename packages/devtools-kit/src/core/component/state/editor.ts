import type { MaybeRef, Ref } from 'vue'
import { isReactive, isRef, toRaw } from 'vue'
import { getComponentInstance } from '../utils'
import { devtoolsContext } from '../../../state'

import type { InspectorStateEditorPayload, PropPath } from '../types'

export type Recordable = Record<PropertyKey, any>

export class StateEditor {
  refEditor = new RefStateEditor()

  set(
    object: Recordable,
    path: PropPath,
    value: unknown,
    cb?: (object: Recordable, field: string, value: unknown) => void,
  ) {
    const sections = Array.isArray(path) ? path : path.split('.')
    const markRef = false
    while (sections.length > 1) {
      const section = sections.shift()!
      object = object[section] as Recordable
      if (this.refEditor.isRef(object))
        object = this.refEditor.get(object)
    }
    const field = sections[0]
    const item = this.refEditor.get(object)[field]
    if (cb) {
      cb(object, field, value)
    }
    else {
      if (this.refEditor.isRef(item))
        this.refEditor.set(item, value)
      else if (markRef)
        object[field] = value
      else
        object[field] = value
    }
  }

  get(object: Recordable, path: PropPath) {
    const sections = Array.isArray(path) ? path : path.split('.')
    for (let i = 0; i < sections.length; i++) {
      object = object[sections[i]] as Recordable
      if (this.refEditor.isRef(object))
        object = this.refEditor.get(object)
      if (!object)
        return undefined
    }
    return object
  }

  has(object: Recordable, path: PropPath, parent = false) {
    if (typeof object === 'undefined')
      return false
    const sections = Array.isArray(path) ? path.slice() : path.split('.')
    const size = !parent ? 1 : 2
    while (object && sections.length > size) {
      const section = sections.shift()!
      object = object[section] as Recordable
      if (this.refEditor.isRef(object))
        object = this.refEditor.get(object)
    }
    return object != null && Object.prototype.hasOwnProperty.call(object, sections[0])
  }

  createDefaultSetCallback(state: InspectorStateEditorPayload['state']) {
    return (object: Recordable, field: string | number, value: unknown) => {
      if (state.remove || state.newKey) {
        if (Array.isArray(object))
          object.splice(field as number, 1)
        else if (toRaw(object) instanceof Map && typeof value === 'object' && value && 'key' in value)
          object.delete(value.key)
        else if (toRaw(object) instanceof Set)
          object.delete(value)
        else
          Reflect.deleteProperty(object, field)
      }
      if (!state.remove) {
        const target = object[state.newKey || field]
        if (this.refEditor.isRef(target))
          this.refEditor.set(target, value)
        else
          object[state.newKey || field] = value
      }
    }
  }
}

class RefStateEditor {
  set(ref: Ref<any>, value: any): void {
    if (isRef(ref)) {
      ref.value = value
    }
    else {
      // if is reactive, then it must be object
      // to prevent loss reactivity, we should assign key by key
      const previousKeys = Object.keys(ref)
      const currentKeys = Object.keys(value)
      // we should check the key diffs, if previous key is the longer
      // then remove the needless keys
      // @TODO: performance optimization
      if (previousKeys.length > currentKeys.length) {
        const diffKeys = previousKeys.filter(key => !currentKeys.includes(key))
        diffKeys.forEach(key => Reflect.deleteProperty(ref, key))
      }
      currentKeys.forEach((key) => {
        Reflect.set(ref, key, Reflect.get(value, key))
      })
    }
  }

  get(ref: MaybeRef<any>): any {
    return isRef(ref) ? ref.value : ref
  }

  isRef(ref: MaybeRef<any>): ref is Ref<any> {
    return isRef(ref) || isReactive(ref)
  }
}

export async function editComponentState(payload: InspectorStateEditorPayload, stateEditor: StateEditor) {
  const { path, nodeId, state, type } = payload
  // assert data types, currently no...
  // if (!['data', 'props', 'computed', 'setup'].includes(dataType))
  const instance = getComponentInstance(devtoolsContext.appRecord!, nodeId)
  if (!instance)
    return

  const targetPath = path.slice()

  let target: Record<string, unknown> | undefined

  // TODO: props
  if (instance.devtoolsRawSetupState && Object.keys(instance.devtoolsRawSetupState).includes(path[0])) {
    // Setup
    target = instance.devtoolsRawSetupState
  }

  if (target && targetPath) {
    if (state.type === 'object' && type === 'reactive') {
      // prevent loss of reactivity
    }
    stateEditor.set(target, targetPath, state.value, stateEditor.createDefaultSetCallback(state))
  }
}

export const stateEditor = new StateEditor()

// TODO: currently (w/ Vue official devtools), we send all state to frontend after changes,
//       we can optimize this by only sending the changed state, to make a better UX.
// PRIORITY: LOW
export async function editState(payload: InspectorStateEditorPayload) {
  editComponentState(payload, stateEditor)
}
