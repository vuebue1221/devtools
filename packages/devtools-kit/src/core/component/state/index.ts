import { devtoolsContext } from '../../../state'
import { getComponentInstance, getInstanceName, getUniqueComponentId } from '../utils'
import { processInstanceState } from './process'

export function getInstanceState(params: { instanceId: string }) {
  const instance = getComponentInstance(devtoolsContext.appRecord!, params.instanceId)
  const id = getUniqueComponentId(instance!)
  const name = getInstanceName(instance!)
  const file = instance?.type?.__file

  const state = processInstanceState(instance!)
  return {
    id,
    name,
    file,
    state,
    instance,
  }
}
