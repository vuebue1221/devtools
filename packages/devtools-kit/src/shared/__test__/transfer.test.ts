import { stringifyReplacer } from '../../core/component/state/replacer'
import { stringifyStrictCircularAutoChunks } from '../transfer'

it('encode', () => {
  const vueInstanceLike = {
    _: {
      ctx: {
        props: {},
      },
      vnode: [] as any[],
      type: {
        props: [],
      },
      appContext: {
        mixins: [],
      },
      setupState: {},
      attrs: {},
      provides: {},
      injects: {},
      refs: {},
    },
    __isVue: true,
    a: 1,
    b: {
      state: [] as any[],
    },
  }

  vueInstanceLike.b.state.push({
    type: 'provided',
    key: '$currentInstance',
    value: vueInstanceLike,
  })

  vueInstanceLike.b.state.push({
    type: 'provided',
    key: '$currentInstance2',
    value: vueInstanceLike,
  })

  vueInstanceLike._.vnode.push(vueInstanceLike)

  expect(stringifyStrictCircularAutoChunks(vueInstanceLike, stringifyReplacer)).toMatchSnapshot()
})
