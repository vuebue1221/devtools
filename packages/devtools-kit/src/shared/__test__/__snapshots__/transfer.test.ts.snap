// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encode 1`] = `" [{"_":1,"__isVue":23,"a":24,"b":25},{"ctx":2,"vnode":4,"type":14,"appContext":16,"setupState":18,"attrs":19,"provides":20,"injects":21,"refs":22},{"props":3},{},[5],{"_custom":6},{"type":7,"id":8,"displayText":9,"tooltipText":10,"value":11,"fields":12},"component","__vue_devtool_undefined__","Anonymous Component","Component instance","__vue_devtool_undefined__",{"abstract":13},true,{"props":15},[],{"mixins":17},[],{},{},{},{},{},true,1,{"state":26},[27,31],{"type":28,"key":29,"value":30},"provided","$currentInstance","[[CircularRef]] <Anonymous Component>",{"type":32,"key":33,"value":34},"provided","$currentInstance2","[[CircularRef]] <Anonymous Component>"]"`;
